### PR TITLE
`Development`: Fix some teamscale findings

### PR DESCRIPTION
**Pull request by MaximilianAnzinger**
#### General

- [x] I chose a title conforming to the [naming conventions for pull requests](https://docs.artemis.cit.tum.de/dev/development-process/#naming-conventions-for-github-pull-requests).

#### Server
- [x] **Important**: I implemented the changes with a very good performance and prevented too many (unnecessary) database calls.
- [x] I **strictly** followed the [server coding and design guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/server/).

#### Performance Review
- [x] I (as a reviewer) confirm that the server changes (in particular related to database calls) are implemented with a very good performance
#### Code Review
- [x] Code Review 1
- [x] Code Review 2

### Test Coverage

unchanged

## Summary by CodeRabbit

- **New Features**
	- Enhanced student score calculation and reporting for improved performance tracking.
- **Documentation**
	- Updated API documentation to include new scoring functionalities and DTO classes.
- **Refactor**
	- Refactored code in ParticipantScoreService to utilize functional interfaces for streamlined data processing.
	- Moved StudentScoresDTO to a new package for better organization.
- **Tests**
	- Improved integration tests in CourseScoreCalculationServiceTest and ParticipantScoreIntegrationTest for robustness.

### CODE REVIEW

File: src/main/java/de/tum/in/www1/artemis/service/ParticipantScoreService.java [hunk 1]
```diff
@@ -117,55 +119,47 @@ private List<ScoreDTO> calculateScores(Set<Exercise> exercises, Set<User> users,
...
         List<Object[]> studentAndAchievedPoints = studentScoreRepository.getAchievedPointsOfStudents(individualExercises);
-        for (Object[] rawData : studentAndAchievedPoints) {
-            User user = (User) rawData[0];
-            double achievedPoints = rawData[1] != null ? ((Number) rawData[1]).doubleValue() : 0.0;
-            if (userIdToScores.containsKey(user.getId())) {
-                userIdToScores.get(user.getId()).pointsAchieved += achievedPoints;
-            }
-        }
+        Map<Long, Double> pointsAchieved = studentAndAchievedPoints.stream().collect(Collectors.toMap(rawData -> ((User) ((Object[]) rawData)[0]).getId(),
+                rawData -> ((Object[]) rawData)[1] != null ? ((Number) ((Object[]) rawData)[1]).doubleValue() : 0D));
```
Comment (Strohgelaender) [inline src/main/java/de/tum/in/www1/artemis/service/ParticipantScoreService.java:133]: Not sure if this really is an improvement, the code is basically unreadable now.

Comment (MaximilianAnzinger) [reply src/main/java/de/tum/in/www1/artemis/service/ParticipantScoreService.java:133]: Please elaborate on which of the concepts are too advanced and might be unclear. 
I'm open to improvements/suggestions.

Comment (MaximilianAnzinger) [reply src/main/java/de/tum/in/www1/artemis/service/ParticipantScoreService.java]: @Strohgelaender @tobias-lippert I've refactored the code. Please let me know if this is now easier to understand.
IMO this is now even easier to read than the previous approach to construct this map.

File: src/main/java/de/tum/in/www1/artemis/repository/StudentScoreRepository.java [hunk 2]
```diff
@@ -42,13 +43,13 @@ public interface StudentScoreRepository extends JpaRepository<StudentScore, Long
 
     // TODO: use a custom object instead of Object[] (as in the example above with ParticipantScoreAverageDTO)
```
Comment (Strohgelaender) [inline src/main/java/de/tum/in/www1/artemis/repository/StudentScoreRepository.java:44]:
```suggestion
```

File: src/main/java/de/tum/in/www1/artemis/service/ParticipantScoreService.java [hunk 3]
```diff
@@ -117,55 +131,46 @@ private List<ScoreDTO> calculateScores(Set<Exercise> exercises, Set<User> users,
...
 
         Course course = exercises.stream().findAny().orElseThrow(() -> new EntityNotFoundException("The result you are referring to does not exist"))
                 .getCourseViaExerciseGroupOrCourseMember();
 
-        // For every student we want to calculate the score
-        Map<Long, ScoreDTO> userIdToScores = users.stream().collect(Collectors.toMap(User::getId, user -> new ScoreDTO(user.getId(), user.getLogin(), 0.0, 0.0, 0.0)));
-
         // individual exercises
         // [0] -> User
         // [1] -> sum of achieved points in exercises
```
Comment (Strohgelaender) [inline src/main/java/de/tum/in/www1/artemis/service/ParticipantScoreService.java:142]:
```suggestion
        // individual exercises
```

**Review by Strohgelaender — APPROVED** [review]

File: src/main/java/de/tum/in/www1/artemis/service/CourseScoreCalculationService.java [hunk 4]
```diff
@@ -13,17 +21,29 @@
...
 import de.tum.in.www1.artemis.service.plagiarism.PlagiarismCaseService.PlagiarismMapping;
-import de.tum.in.www1.artemis.web.rest.dto.*;
+import de.tum.in.www1.artemis.web.rest.dto.BonusSourceResultDTO;
+import de.tum.in.www1.artemis.web.rest.dto.CourseForDashboardDTO;
+import de.tum.in.www1.artemis.web.rest.dto.CourseScoresDTO;
+import de.tum.in.www1.artemis.web.rest.dto.ParticipationResultDTO;
+import de.tum.in.www1.artemis.web.rest.dto.score.StudentScoresDTO;
 
 /**
  * Service Implementation for calculating course scores.
```
[the remark this answers is not in the document]
Comment (Strohgelaender) [reply src/main/java/de/tum/in/www1/artemis/service/CourseScoreCalculationService.java:49]: After a discussion within the team, we decided to consistently not use star-imports in the project. In the future, please don't add review comments about star imports.

**Review by tobias-lippert — APPROVED** [review]

**Pull request merged**